storm-core/src/jvm/org/apache/storm/pacemaker/Pacemaker.java

```diff
@@ -0,0 +1,246 @@
+            }
+            HBPulse hbPulse = new HBPulse();
+            hbPulse.set_id(path);
+            hbPulse.set_details(details);
+            return new HBMessage(HBServerMessageType.GET_PULSE_RESPONSE, HBMessageData.pulse(hbPulse));
+        } else {
+            return notAuthorized();
+        }
+    }
+
+    private HBMessage deletePath(String path) {
+        String prefix = path.endsWith("/") ? path : (path + "/");
+        for (String key : heartbeats.keySet()) {
+            if (key.indexOf(prefix) == 0)
+                deletePulseId(key);
+        }
+        return new HBMessage(HBServerMessageType.DELETE_PATH_RESPONSE, null);
+    }
+
+    private HBMessage deletePulseId(String path) {
+        LOG.debug("Deleting Pulse for id [ {} ].", path);
+        heartbeats.remove(path);
+        return new HBMessage(HBServerMessageType.DELETE_PULSE_ID_RESPONSE, null);
+    }
+
+    private void updateLargestHbSize(int size) {
+        int newValue = size;
+        while (true) {
+            int oldValue = pacemakerStats.largestHeartbeatSize.get();
+            if (newValue > oldValue) {
+                if (!pacemakerStats.largestHeartbeatSize.compareAndSet(oldValue, newValue))
+                    continue;
+            }
+            break;
+        }
+    }
+
+    private void updateAverageHbSize(int size) {
+        while (true) {
+            int oldValue = pacemakerStats.averageHeartbeatSize.get();
+            int count = pacemakerStats.sendPulseCount.get();
+            int newValue = ((count * oldValue) + size) / (count + 1);
+            if (!pacemakerStats.averageHeartbeatSize.compareAndSet(oldValue, newValue))
+                continue;
+            break;
+        }
+    }
+
+    private void startStatsThread() {
+        Callable afn = new Callable() {
+            public Object call() {
+                int sendCount = pacemakerStats.sendPulseCount.getAndSet(0);
+                int receivedSize = pacemakerStats.totalReceivedSize.getAndSet(0);
+                int getCount = pacemakerStats.getPulseCount.getAndSet(0);
+                int sentSize = pacemakerStats.totalSentSize.getAndSet(0);
+                int largest = pacemakerStats.largestHeartbeatSize.getAndSet(0);
+                int average = pacemakerStats.averageHeartbeatSize.getAndSet(0);
+                int totalKeys = heartbeats.size();
+                LOG.debug(
+                        "\nReceived {} heartbeats totaling {} bytes,\nSent {} heartbeats totaling {} bytes,\nThe largest heartbeat was {} bytes,\nThe average heartbeat was {} bytes,\nPacemaker contained {} total keys\nin the last {} second(s)",
+                        sendCount, receivedSize, getCount, sentSize, largest, average, totalKeys, sleepSeconds);
+                return sleepSeconds; // Run only once.
+            }
+        };
+        Utils.asyncLoop(afn, true, null, Thread.currentThread().getPriority(), false, true, null);
```

Comment:
it will be nice to replace true, false with variables just to know what are being passed, but it is upto you